python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll

```diff
@@ -447,7 +447,87 @@ class LocalSourceNode extends Node {
 
   /** Holds if this `LocalSourceNode` can flow to `nodeTo` in one or more local flow steps. */
   cached
-  predicate flowsTo(Node nodeTo) { simpleLocalFlowStep*(this, nodeTo) }
+  predicate flowsTo(Node nodeTo) { Cached::hasLocalSource(nodeTo, this) }
+
+  /**
+   * Gets a reference (read or write) of attribute `attrName` on this node.
+   */
+  AttrRef getAnAttributeReference(string attrName) { Cached::namedAttrRef(this, attrName, result) }
+
+  /**
+   * Gets a read of attribute `attrName` on this node.
+   */
+  AttrRead getAnAttributeRead(string attrName) { result = getAnAttributeReference(attrName) }
+
+  /**
+   * Gets a reference (read or write) of any attribute on this node.
+   */
+  AttrRef getAnAttributeReference() {
+    Cached::namedAttrRef(this, _, result)
+    or
+    Cached::dynamicAttrRef(this, result)
+  }
+
+  /**
+   * Gets a read of any attribute on this node.
+   */
+  AttrRead getAnAttributeRead() { result = getAnAttributeReference() }
+
+  /**
+   * Gets an invocation (with our without `new`) of this node.
+   */
+  Node getAnInvocation() { Cached::invocation(this, result) }
```

Comment:
Could we use `getACall` instead? that way it fits with our old (points-to based) terminology, and the terminology introduced in API graphs :blush: 

I'm slightly afraid to ask this knowing that we're (long term) aiming to be able to share this between different languages, but it just looked to strange to see _invocation_ used instead of _call_

Reply:
Yeah, this one should probably just reuse the `getACall` terminology. Also, whoops, I notice that the comment mentions `new`. A leftover from the JS libs. :sweat_smile: